### PR TITLE
refactor: inject dependency `zm_searchUserCache`

### PR DESCRIPTION
**Pull request by markusfassbender**
### Issue

Objective: Remove dependency `zm_searchUserCache` that was made available through `NSManagedObjectContext`.

It seems that it was not used a lot, but because of this implicit dependency via contexts it has big impact.

Screenshot 2024-04-25 at 12 15 55

### Changes

> [!NOTE]
> First review `UserSession` and `ZMUserSession` to see the important changes.

- Add new `UserSessionDependencies` to retain dependencies. This can be evolved to a core technology in future.
- Add `typealias SearchUsersCache` to highest possible module `WireDataModel`
- Update some files to inject the cache.
- Update many unit and snapshot tests.
- Remove wrong values from git ignore

### Testing

- TBD: ?

---

### Checklist

- [ ] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

### CODE REVIEW

File: wire-ios-data-model/Tests/Source/Model/Observer/SearchUserObserverCenterTests.swift [hunk 1]
```diff
@@ -18,184 +18,13 @@
 
 @testable import WireDataModel
 
-class SearchUserSnapshotTests: ZMBaseManagedObjectTest {
```
Comment (markusfassbender) [inline wire-ios-data-model/Tests/Source/Model/Observer/SearchUserObserverCenterTests.swift:21]: There were two classes in one file, so I've moved it to a new file `SearchUserSnapshotTests.swift`

File: wire-ios-sync-engine/Source/UserSession/ZMUserSession/Dependencies/UserSessionDependencies.swift [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
```
Comment (caldrian) [inline wire-ios-sync-engine/Source/UserSession/ZMUserSession/Dependencies/UserSessionDependencies.swift]: Nothing used from `Foundation` in this file.

File: wire-ios-sync-engine/Source/UserSession/ZMUserSession/Dependencies/Caches.swift [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
```
Comment (caldrian) [inline wire-ios-sync-engine/Source/UserSession/ZMUserSession/Dependencies/Caches.swift]: Why is `Foundation` needed?

**Review by caldrian — APPROVED** [review]
That's one small step for a man ... 🧑‍🚀
Please fix the SwiftLint warning.

File: wire-ios-sync-engine/Source/Synchronization/Strategies/SearchUserImageStrategy.swift [hunk 4]
```diff
@@ -32,17 +32,24 @@ public class SearchUserImageStrategy: AbstractRequestStrategy {
     fileprivate var requestedPreviewAssetsInProgress: Set<UUID> = Set()
     fileprivate var requestedCompleteAssetsInProgress: Set<UUID> = Set()
 
-    fileprivate var observers: [Any] = []
+    fileprivate var observers: [any NSObjectProtocol] = []
+
+    private let searchUsersCache: SearchUsersCache?
```
Comment (netbe) [inline wire-ios-sync-engine/Source/Synchronization/Strategies/SearchUserImageStrategy.swift:37]: is it supposed to be optional?

File: wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession.swift [hunk 5]
```diff
@@ -499,11 +504,12 @@ public final class ZMUserSession: NSObject {
         let fileAssetCache = FileAssetCache(location: cacheLocation)
         let userImageCache = UserImageLocalCache(location: cacheLocation)
 
-        self.assetCache = fileAssetCache
+        dependencies.caches.fileAssets = fileAssetCache
+        dependencies.caches.userImages = userImageCache
+        dependencies.caches.searchUsers = NSCache()
```
Comment (netbe) [inline wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession.swift]: if we're going to move deps to own object, shouldn't it be the responsability of UserSessionDependencies to create those Cache objects ?

File: wire-ios-sync-engine/Source/Synchronization/Strategies/SearchUserImageStrategy.swift (hunk 4 above)
Comment (markusfassbender) [reply wire-ios-sync-engine/Source/Synchronization/Strategies/SearchUserImageStrategy.swift:37]: In this case I tried not to change too much of the existing parts, but after our call I could do the dependency object non-optional: 985f0728dce45937f14b57eee37825510e7d2c7a

File: wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession.swift (hunk 5 above)
Comment (markusfassbender) [reply wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession.swift]: Good point, was moved to the builder: f3dbf54a962f77959b430462d39c0f607816c1b6

File: wire-ios-mono.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 6]
```diff
@@ -1,24 +1,6 @@
 {
   "object": {
     "pins": [
-      {
```
Comment (netbe) [inline wire-ios-mono.xcworkspace/xcshareddata/swiftpm/Package.resolved]: should this be removed ?

**Review by netbe — APPROVED** [review]
Good work 👏

File: wire-ios-mono.xcworkspace/xcshareddata/swiftpm/Package.resolved (hunk 6 above)
Comment (markusfassbender) [reply wire-ios-mono.xcworkspace/xcshareddata/swiftpm/Package.resolved]: this is a mistake, thanks! 3b0cd5755d61bb3437b49779492ab848d59e1148

**Pull request merged**